rhcos/jobspec.yaml

```diff
@@ -0,0 +1,102 @@
+      # https://issues.redhat.com/browse/ART-2355
+      exclude=nss-altfiles kernel protobuf
+      # This repo provides 'qemu-kiwi' for the `sandboxed-containers` extension
+      # https://github.com/openshift/enhancements/blob/master/enhancements/sandboxed-containers/sandboxed-containers-tech-preview.md
+
+      [rhel-8-advanced-virt]
+      baseurl=http://rhsm-pulp.corp.redhat.com/content/dist/layered/rhel8/$basearch/advanced-virt/os/
+      enabled=1
+      gpgcheck=0
+      [rhel-8-fast-datapath]
+      baseurl=http://rhsm-pulp.corp.redhat.com/content/dist/layered/rhel8/$basearch/fast-datapath/os/
+      enabled=1
+      gpgcheck=0
+
+      # Fast Datapath has yet to ship for ARM, so the above repo is empty on aarch64.
+      # Once it starts shipping, the following repo should be removed.
+      [rhel-8-fast-datapath-aarch64]
+      baseurl=http://download-node-02.eng.bos.redhat.com/rcm-guest/puddles/Multi-Arch/Fast-DataPath/EL8-Production/latest/aarch64/os/
+      enabled=1
+      gpgcheck=0
+      # The baseurls below are tracking the latest version of RHEL that has been released.
+      # In the case of any of these locations being unavailable, it should be possible
+      # to switch out the `/8/` to the latest RHEL version (i.e. `/8.1/` or `/8.2/`).
+      #
+      # RHCOS 4.8 will track RHEL 8.4, but we can use `8` until 8.4 reaches EUS
+      [rhel-8-baseos]
+      baseurl=http://rhsm-pulp.corp.redhat.com/content/dist/rhel8/8/$basearch/baseos/os/
+      enabled=1
+      gpgcheck=0
+
+      [rhel-8-appstream]
+      baseurl=http://rhsm-pulp.corp.redhat.com/content/dist/rhel8/8/$basearch/appstream/os/
+      enabled=1
+      gpgcheck=0
+      # Exclude toolbox to make sure we use coreos/toolbox from rhaos repo
+      # kata-containers wants non-modular libpmem
+      exclude=toolbox libpmem*module*
+
+      [rhel-8-nfv]
+      baseurl=http://rhsm-pulp.corp.redhat.com/content/dist/rhel8/8/x86_64/nfv/os/
+      enabled=1
+
+publish_oscontainer:
+  buildstrategy_tls_verify: null
+  registries:
+  - url: quay.io/openshift-release-dev/ocp-v4.0-art-dev
+    secret_type: cluster
+    secret: upload-registry.ci.openshift.org.dockercfg
+  - url: registry.ci.openshift.org/rhcos/machine-os-content:test
+    # should be :4.y
+    secret_type: cluster
+    secret: upload-registry.ci.openshift.org.dockercfg
+
+stages:
+- id: Build and test QEMU
+  description: Create the base build and test it
+  build_artifacts: [base]
+  post_commands:
+    - cosa kola run --basic-qemu-scenarios
+  execution_order: 1
+- id: Upload to brew
+  description: Create brew build record and upload artifacts
+  require_artifacts: [qemu]
+  commands:
+    - cosa ... ?
+      # secret is rhcos-keytab
+      # can this retry on failure?
+  execution_order: 2
+- id: Upload to S3
+  description: Insert into S3 (but not in standard stream location)
+  require_artifacts: [qemu]
+  commands:
+    - cosa buildupload s3 --acl=private art-rhcos-ci/testing/gangplank
```

Comment:
Before running buildupload we need to check for previous builds
`coreos-assembler buildprep s3//url`
```
- id: Build Prep
  description: ' Check previous build'
  require_artifacts: [qemu]
  post_commands: [coreos-assembler buildprep s3://art-rhcos-ci/releases/rhcos-4.9]
  execution_order: 3
```

Can you check if it will work?

Reply:
running now at https://privileged.psi.redhat.com/console/project/rhcos-art/browse/builds/cosa-runner-main/cosa-runner-main-11?tab=logs

I'm wondering though if this could just be another command that runs in the next stage before `cosa buildupload s3 ...`?

And also if running this parallel to brew would work

Reply:
logs weren't retained from that one, started https://privileged.psi.redhat.com/console/project/rhcos-art/browse/builds/cosa-runner-main/cosa-runner-main-12?tab=logs

Reply:
```


time="2021-09-01T12:55:56Z" level=info msg="No build history found" require_artifacts="[qemu]" stage="Upload to brew"
--
  | time="2021-09-01T13:55:56Z" level=error msg="Stage sent error" error="required artifacts never appeared"
  | time="2021-09-01T13:55:56Z" level=info msg="logging terminated" logrus_error="can not add field \"logfile\"" container=init pod=cosa-runner-main-12-worker-0
  | time="2021-09-01T13:55:56Z" level=info msg="No build history found" require_artifacts="[qemu]" stage="Build Prep"
  | time="2021-09-01T14:55:56Z" level=error msg="Stage sent error" error="required artifacts never appeared"
  | time="2021-09-01T14:55:56Z" level=info msg="logging terminated" logrus_error="can not add field \"logfile\"" container=cosa-runner-main-12-worker-0 pod=cosa-runner-main-12-worker-0
  | time="2021-09-01T14:55:56Z" level=info msg="No build history found" require_artifacts="[qemu]" stage="Upload to S3"
```

also... shouldn't earlier stage failures result in later stages not running?

Reply:
i finally realized my cleanup script was deleting the errored build pods. started [another](https://privileged.psi.redhat.com/console/project/rhcos-art/browse/builds/cosa-runner-main/cosa-runner-main-13?tab=logs)...

Reply:
> https://privileged.psi.redhat.com/console/project/rhcos-art/browse/builds/cosa-runner-main/cosa-runner-main-11?tab=logs

Hi Luke, Looks I accidentally stopped notifications for this PR my bad for the late response here.
Yeah you could run it in another stage, you just need to make sure it will follow the prep build and then the upload.
Brew doesn't use this prep build. I can't see any issue running it in parallel if that's your question.

Reply:
The logs are no longer there. What is the current state for it?

Reply:
Another run: https://privileged.psi.redhat.com/console/project/rhcos-art/browse/builds/cosa-runner-main/cosa-runner-main-14?tab=logs

Reply:
umm brew does make use of the meta.json, we are probably missing `delay_meta_merge: true` or even missing the `finalize` stage. Let me run it here to try to understand/validate it.

Reply:
@sosiouxme Can you try to add the finalize stage after the build base and also add  `delay_meta_merge: true`? In my tests I did it:

```
...
- id: Finalize Build
  description: Return meta.json updated
  build_artifacts: [finalize]
  execution_order: 2
- id: Upload to brew
  description: Create brew build record and upload artifacts
  require_artifacts: [qemu]
  commands: [coreos-assembler koji-upload upload --reserve-id --keytab /srv/.keytab/keytab
      --buildroot builds --owner rhcos-build/jenkins-redhat-coreos.cloud.paas.upshift.redhat.com@REDHAT.COM
      --profile brew --tag rhaos-4.9-rhel-8-build]
  execution_order: 3
delay_meta_merge: true
```
```
DEBU[0772] all dependencies for stage have been meet     require_artifacts="[qemu]" stage="Upload to brew"
INFO[0772] Worker dependences have been defined          require_artifacts="[qemu]" stage="Upload to brew"
INFO[0772] Executing worker pod                          require_artifacts="[qemu]" stage="Upload to brew"
INFO[0772] Creating pod cosa-01a24be4-6afb-4d46-aec1-16d46ad1c4de-worker-2
INFO[0772] InitContainer has been defined
INFO[0772] Pod created: cosa-01a24be4-6afb-4d46-aec1-16d46ad1c4de-worker-2
os-4.9-rhel-8-build
...
+ coreos-assembler koji-upload upload --reserve-id --keytab /srv/.keytab/keytab --buildroot builds --owner rhcos-build/jenkins-redhat-coreos.cloud.paas.upshift.redhat.com@REDHAT.COM --profile brew --tag rhaos-4.9-rhel-8-build
cosa-01a24be4-6afb-4d46-aec1-16d46ad1c4de-worker-2 [+6.994s]: [INFO]: Targeting architecture: ppc64le
cosa-01a24be4-6afb-4d46-aec1-16d46ad1c4de-worker-2 [+6.994s]: [INFO]: Targeting build: 34.20210913.dev.0

```

However I'm getting:
```FileNotFoundError: [Errno 2] No such file or directory: 'builds/34.20210913.dev.0/ppc64le/coreos-assembler-config-git.json'```

The RHCOs CI is created using /home/builder/builds, when gangplank uses builder dir as default, since no build was created changing it to `--bucket builds` ( dir/builds) is not working, gangplank still creates a builder dir inside the builds. I'm checking how we can make this work now.

Reply:
Just saw it now: ` --buildroot builder` in ` coreos-assembler koji-upload upload ` may work

Reply:
Actually gangpplank is not returning the `coreos-assembler-config-git.json'`, I'm testing a fix for it.

Reply:
https://privileged.psi.redhat.com/console/project/rhcos-art/browse/builds/cosa-runner-main/cosa-runner-main-15?tab=logs to at least try out your suggestion (latest commit), sorry for inattention

Reply:
@ravanelli guessing this might just be because there are no builds in the stream already (no stream exists)?
```


cosa-runner-main-15-worker-1 [+597ms]: + cosa meta --finalize
--
  | cosa-runner-main-15-worker-1 [+765ms]: Traceback (most recent call last):
  | cosa-runner-main-15-worker-1 [+765ms]:   File "/usr/lib/coreos-assembler/cmd-meta", line 128, in <module>
  | cosa-runner-main-15-worker-1 [+765ms]:     new_cli()
  | cosa-runner-main-15-worker-1 [+765ms]:   File "/usr/lib/coreos-assembler/cmd-meta", line 63, in new_cli
  | cosa-runner-main-15-worker-1 [+765ms]:     meta = GenericBuildMeta(args.workdir, args.build, basearch=args.arch, schema=schema)
  | cosa-runner-main-15-worker-1 [+767ms]:   File "/usr/lib/coreos-assembler/cosalib/meta.py", line 266, in __init__
  | cosa-runner-main-15-worker-1 [+767ms]:     build = builds.get_latest()
  | cosa-runner-main-15-worker-1 [+767ms]:   File "/usr/lib/coreos-assembler/cosalib/builds.py", line 70, in get_latest
  | cosa-runner-main-15-worker-1 [+769ms]:     return self._data['builds'][0]['id']
  | cosa-runner-main-15-worker-1 [+769ms]: IndexError: list index out of range
  | cosa-runner-main-15-worker-1 [+791ms]: time="2021-09-16T18:10:48Z" level=error msg="failed stage execution" build artifacts="[finalize]" commands="[]" error="Script exited with return code exit status 1" optional

```

Reply:
Yeah, not sure why. The qemu went well, but we got an error with the kola tests that didn't finish well. Maybe that's the reason the artifacts weren't add in the finalize part. Can you trigger it again?

Reply:
https://privileged.psi.redhat.com/console/project/rhcos-art/browse/builds/cosa-runner-main/cosa-runner-main-16?tab=logs

Reply:
```
cosa-runner-main-16-worker-0 [+8m13.526s]: + cosa kola run --basic-qemu-scenarios
--
  | cosa-runner-main-16-worker-0 [+8m13.692s]: kola --output-dir tmp/kola run --qemu-nvme=true basic
  | cosa-runner-main-16-worker-0 [+8m13.73s]: ⚠️  Skipping kola test pattern "coreos.ignition.journald-log":
  | cosa-runner-main-16-worker-0 [+8m13.73s]:   ߑ頨ttps://github.com/coreos/coreos-assembler/issues/1173
  | cosa-runner-main-16-worker-0 [+8m13.738s]: === RUN   basic
  | cosa-runner-main-16-worker-0 [+8m13.743s]: qemu-img: /var/tmp/mantle-qemu372520528/disk022282607: Image creation needs a size parameter
  | cosa-runner-main-16-worker-0 [+8m13.745s]: --- FAIL: basic (0.01s)
  | cosa-runner-main-16-worker-0 [+8m13.745s]:         harness.go:1124: Cluster failed starting machines: exit status 1
  | cosa-runner-main-16-worker-0 [+8m13.745s]: FAIL, output in tmp/kola
  | cosa-runner-main-16-worker-0 [+8m13.745s]: Error: harness: test suite failed
  | cosa-runner-main-16-worker-0 [+8m13.745s]: 2021-09-17T18:11:33Z cli: harness: test suite failed
  | cosa-runner-main-16-worker-0 [+8m13.748s]: Traceback (most recent call last):
  | cosa-runner-main-16-worker-0 [+8m13.748s]:   File "/usr/lib/coreos-assembler/cmd-kola", line 65, in <module>
  | cosa-runner-main-16-worker-0 [+8m13.748s]:     subprocess.check_call(subargs)
  | cosa-runner-main-16-worker-0 [+8m13.748s]:   File "/usr/lib64/python3.9/subprocess.py", line 373, in check_call
  | cosa-runner-main-16-worker-0 [+8m13.748s]:     raise CalledProcessError(retcode, cmd)
  | cosa-runner-main-16-worker-0 [+8m13.748s]: subprocess.CalledProcessError: Command '['kola', '--output-dir', 'tmp/kola', 'run', '--qemu-nvme=true', 'basic']' returned non-zero exit status 1.
  | cosa-runner-main-16-worker-0 [+8m13.776s]: time="2021-09-17T18:11:33Z" level=error msg="failed stage execution" build artifacts="[base]" commands="[]" error="Script exited with return code exit status 1" optional artifacts="[]" required artifacts="[]" return files="[]" stage id="Build and test QEMU"
```

Reply:
What is interesting is, it is able to get the qemu artifacts done, it also starts the upload to minio: 
```
 level=info msg=Uploaded bucket=builder etag=98832972d2b7e77b404ec6abcf4eda00-156 from=/srv/builds/49.84.202109171805-0/x86_64/rhcos-49.84.202109171805-0-qemu.x86_64.qcow2 func=putter gangplank.coreos.com/cosa/stamp=1631902331248630933 object=49.84.202109171805-0/x86_64/rhcos-49.84.202109171805-0-qemu.x86_64.qcow2 remote size=2613706752 size=2613706752
```
And we get an error in the end, I will need to debug it to understand if the upload for minio was completed ok indeed. 
```
  | time="2021-09-17T18:12:12Z" level=error msg="Stage sent error" error="Build and test QEMU failed: container cosa-runner-main-16-worker-0 exited with code 1"
```
And for the new pod created for finalize, indeed no build history is found, so  qemu is never uploaded to this pod. 

```
time="2021-09-17T18:12:12Z" level=info msg="No build history found" require_artifacts="[]" stage="Finalize Build"
```

In the local tested I did, it worked well.

```
Found new build ID" buildID=34.20210920.dev.0
level=debug msg="will request" bucket=builder buildID=34.20210920.dev.0 object...
...
 cosa meta --workdir /srv --build 34.20210920.dev.0 --dump
```
One thing we can do to just to try to validate the rest and try to understand more, is to created a new stage for the kola test,  instead of running it as pos commands. 

```
stages:
- id: Test kola
  description: 'Stage 2 execution '
  require_artifacts: [qemu]
  post_commands: [cosa kola --basic-qemu-scenarios]
  execution_order: 2

```